### DISCUSSION TITLE
inheriting from a "parent" machine

### DISCUSSION THREAD

**Discussion by leskat47** [statelyai/xstate#2612] [Q&A]
Hi all. I'm new to xstate and I'm trying to set up state machines that are all derivative of one machine.

Is there a way to have a "parent" machine from which others can be based with variations, like perhaps some additional states and actions? Is `withConfig` the correct way to go?

Appreciate the help.

**Comment by ChrisShank** [comment]
Hi! Xstate supports the ability to swap out the initial context using `createMachine(...).withContext(...)` and swapping out actions, guards, and services using `createMachine(...).withConfig(...)`. In general I find that using a factory function that returns a machine is a little more robust since its shows the intent of what is needed for the machine to run and is more typesafe! For example lets say that you have a fetching machine (maybe it models some retry logic) that you want to reuse with different fetcher:

```ts
function createFetcher<Data>(fetcher: () => Promise<Data>) {
  return createMachine(
    {
      id: 'generic fetcher',
      initial: 'idle',
      states: {
        idle: {},
        loading: {
          invoke: {
            src: 'fetcher',
            onDone: 'done'
          }
        },
        done: {},
      },
    },
    {
      services: {
        fetcher: () => fetcher()
      }
    },
   
  )
}
```
